odoo/index.py

```diff
@@ -0,0 +1,285 @@
+        return (
+            f'create index {maybe_quote(index_name)} '
+            f'on {maybe_quote(schema)}.{maybe_quote(table)} '
+            f'using {self.method} '
+            f'({expression})'
+            f'{where}'
+        ).format(field=field.name)
+
+    def expression(self, env, field):
+        if field.translate:
+            _logger.warning(
+                "Using a non-trigram index on translated field %s is "
+                "useless, ignoring.",
+                field
+            )
+            return None
+
+        return maybe_quote(field.name)
+
+
+@dataclasses.dataclass
+class unique(index):
+    nulls: Nulls = Nulls.DISTINCT
+    # error message on constraint failure
+    message: str = ''
+
+    def serialize(self, env, field):
+        query = super().serialize(env, field)
+        if query is None:
+            return None
+
+        return query.replace('create index', 'create unique index')
+
+
+@dataclasses.dataclass
+class btree(index):
+    pass
+
+# maybe should be class methods of btree?
+def not_null(**kw) -> btree:
+    return btree(**kw, where="{field} is not null")
+
+@dataclasses.dataclass
+class _boolean(index):
+    def expression(self, env, field):
+        if field.type != 'boolean':
+            _logger.warning(
+                "Boolean partial indexes should only be used on boolean "
+                "fields, %s (%s) is incorrect", field, field.type)
+            return
+        # not useful to index the value we're specifically filtering on, so
+        # index id instead to get a covering index
+        return 'id'
+
+@dataclasses.dataclass
+class true(_boolean):
+    where: str = '{field}'
+
+@dataclasses.dataclass
+class false(_boolean):
+    where: str = 'not {field}'
+
+@dataclasses.dataclass
+class hash(index): # pylint: disable=redefined-builtin
+    method: str = 'hash'
+
+@dataclasses.dataclass
+class trigram(index):
+    method: str = 'gin'
+    def expression(self, env, field):
+        expr = maybe_quote(field.name)
+        text = field.type == 'text'
+        if field.translate:
+            expr = f"jsonb_path_query_array({expr}, '$.*'::jsonpath)"
+            text = False
+        if not text:
+            expr = f'({expr})::text'
+        if hasattr(field, 'unaccent') and field.unaccent:
+            if env.registry.has_unaccent == FunctionStatus.INDEXABLE:
+                expr = get_unaccent_wrapper(env.cr)(expr)
+            else:
+                if env.registry.has_unaccent == FunctionStatus.PRESENT:
+                    warnings.warn("Unaccent present but not immutable")
+                expr = f'({expr})'
+        else:
+            expr = f'({expr})'
+
+        # there's actually a failure point: if `field` is compatible with
+        # gin_trgm_ops (which is?) and neither function gets applied, then
+        # there are no wrapper parens or type casting
+        return f'{expr} gin_trgm_ops'
+
+
+STRING_TO_TYPE: Dict[str, index] = {
```

Comment:
Name is confusing: values are not types by instances.  I suggest `STRING_TO_INDEX`, which is more explicit and avoids the ambiguity of the word "type" (Python type vs index type).